src/splitwise.js

```diff
@@ -614,28 +627,70 @@ class Splitwise {
       logger({ level: LOG_LEVELS.ERROR, message });
       throw new Error(message);
     }
-
+    // check if no redirect url supplied to isAuthorizationCode
+    if (isAuthorizationCode && !redirect_uri) {
+      const message = 'Redirect url required for OAuth2';
+      logger({ level: LOG_LEVELS.ERROR, message });
+      throw new Error(message);
+    }
+    
     const oauth2 = new OAuth2(
       consumerKey,
       consumerSecret,
-      'https://secure.splitwise.com/',
-      null,
-      'oauth/token',
+      SPLITWISE_ENDPOINTS.path,
+      isAuthorizationCode ? SPLITWISE_ENDPOINTS.authorizePath : null,
+      SPLITWISE_ENDPOINTS.accessTokenPath,
       null
     );
 
-    const accessTokenPromise = (() => {
+    this.getAuthorizationUrl = () => {
+      if (!isAuthorizationCode) return "";
+      return oauth2.getAuthorizeUrl({
+        redirect_uri,
+        scope: '',
+        state,
+        response_type: 'code'
+      });
+    }
+
+    const getAccessTokenFromAuthCode = () => {
+      return new Promise((resolve, reject) => {
+        if(!this.authCode) 
+          return reject(`No auth code generated yet. Visit ${this.getAuthorizationUrl()} and login. You need a callback url in your project as mentioned in callback url in your registered splitwise app. Then call \`getAccessToken()\` to register the access token.`);
+        if(this.accessToken) return resolve(this.accessToken);
```

Comment:
Ah sorry, I realize now this might not work either. What we have to remember is that with the `client_credentials` flow, the client will be receiving authorization on behalf of exactly one person (the person owning the credentials) so it makes sense to store data on `this`, however in the `authorization_code` flow, we have 1 client <=> 1 instance of this library <=> many users, which means that there isn't just one access token, there are many. This line here would give away the access token of the first user who authenticates to whoever comes next, a security problem! In general, we should probably entirely avoid using `this` for any code that handles `authorization_code` flow

Reply:
@keriwarr Yaa I thought about this. With this, our clients will need to use the `authorization_code` to allow multiple users interaction. Because, with `client_credentials`, client can just use his/her account only. As it logs into his own account be default. So `authorization_code` would become mandatory in case of multiple users. 
Like wise, talking about the state, it should be supplied in `getAuthorizationUrl` and then once logged in, we would give away the access token and we wont have track of it at all as multiple users would be interacting with the client.
So, summing up,
1. We can remove the use of this but we wont have the access_token and then the client will have to supply the access_token for each function. Not a good idea
2. Client can generate and share the state and we wont have to deal with it at all. Better for both, client as well as the library.
3. Else, to track the state, we can have a redis connection to persist the state. But to which user does it belong to?


Also, as per you, will this library be used for a project of multiple users?